### ISSUE TITLE
Trouble building on a mac

### ISSUE THREAD

**Issue by wsnoble** [crux-toolkit/crux-toolkit#627]
(This is from a crux user...)

I am trying to build the source version of Crux (v4.1.decd99ff that I have been using throughout the paper) on my mac m1 but having trouble. I am even having some trouble doing it on my intel-based mac laptop (I was previously running Crux via Linux). I get the following error on my intel-laptop:

```
builtins.c:490:5: error: implicit declaration of function 'init_svnrev' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    init_svnrev();
    ^
1 error generated.
Building pwiz...
./quickbuild.sh: line 36: /Users/jackfreestone/Downloads/crux-toolkit/ext/build/src/ProteoWizard/libraries/boost-build/src/engine/bin/Darwin/bjam: cannot execute binary file
At least one pwiz target failed to build.
CMake Error at /Users/jackfreestone/Downloads/crux-toolkit/ext/build_pwiz.cmake:18 (message):
  ProteoWizard build failed
Call Stack (most recent call first):
  /Users/jackfreestone/Downloads/crux-toolkit/ext/build_pwiz.cmake:83 (check_status)
```

I have looked online and I think implicit declaration of function triggers errors instead of warnings now as default? I figure there is a way to resolve this but I know nothing about C and compilers. Any help would be greatly appreciated!

**Comment by freejstone** [comment]
Commenting so I can follow this.

**Comment by CharlesEGrant** [comment]
The error is in the build of ProteoWizard which is one of our external dependencies. Since it is an external project we have very limited (usually non-existent) influence over their build.

The underlying problem is that each new release of macOS involves a new release of Apple's development environment (XCode and associated compilers and tools). Each new release of the compiler generally becomes stricter in its interpretation of the language, and often requires changes to existing source in order to compile. This is particularly true when it involves a switch to an entirely new processor architecture like the Intel->M1 transition.

Is there a reason you are building from source? If you have copies of the Crux binary you built previously, these should still work on new versions of the OS. However, on the M1, binaries built on Intel will run in emulation mode and will be significantly slower. 

If you do need to build from source, you'll need to upgrade to a newer version of the software. I can confirm that the latest daily release does build on the macOS for both Intel and M1.

Finally, if you need to build from source, and absolutely must stay with the decd99ff release, you'll have to download and install an older version of the Apple development tools from the Apple developer web site. I'm not sure this is an option for the M1.

**Comment by freejstone** [comment]
Hmm I have tried using an older version of Xcode command line tools (v13) and the corresponding clang compiler that comes with it (13.0.0) to install the `decd99f` release on two intel-based macs and I appear to still get the same error on both. Unfortunately I can't resort to an even older version of Xcode as I would need an older macOS. I've also tried using an older gnu compiler (gnu@7 in homebrew). I haven't reverted cmake, make, git, gnutar, so maybe I need to do this as well just to be sure (though it doesn't seem like they should matter).

I might try and see if I can build it on a Windows virtual machine on my M1 mac. I essentially would like to run the `decd99f` release on the M1 -- so if I could build it on an intel-based mac and copy the binary after build to the M1 or have it built on the M1 directly somehow, then goal achieved.

Thanks for your help!
